include/RAJA/exec-cuda/reduce_cuda.hxx

```diff
@@ -931,104 +955,111 @@ class ReduceMinLoc<cuda_reduce<BLOCK_SIZE>, T> {
     __shared__ CudaReductionLocBlockDataType sd[BLOCK_SIZE];
     __shared__ bool lastBlock;
 
+    int blockId = blockIdx.x 
+       + blockIdx.y * gridDim.x 
+       + gridDim.x * gridDim.y * blockIdx.z; 
+    int threadId = threadIdx.x + 
+               blockDim.x * threadIdx.y + 
+               (blockDim.x * blockDim.y) * threadIdx.z;
     // initialize shared memory
     for (int i = BLOCK_SIZE / 2; i > 0; i /= 2) {
       // this descends all the way to 1
-      if (threadIdx.x < i) {
+      if (threadId < i) {
         // no need for __syncthreads()
-        sd[threadIdx.x + i].val = m_reduced_val;
-        sd[threadIdx.x + i].idx = m_reduced_idx;
+        sd[threadId + i].val = m_reduced_val;
+        sd[threadId + i].idx = m_reduced_idx;
       }
     }
     __syncthreads();
 
-    sd[threadIdx.x].val = val;
-    sd[threadIdx.x].idx = idx;  // need to reconcile loc vs idx naming
+    sd[threadId].val = val;
+    sd[threadId].idx = idx; 
     __syncthreads();
 
     for (int i = BLOCK_SIZE / 2; i >= WARP_SIZE; i /= 2) {
-      if (threadIdx.x < i) {
-        sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x], sd[threadIdx.x + i]);
+      if (threadId < i) {
+        sd[threadId] = RAJA_MINLOC(sd[threadId], sd[threadId + i]);
       }
       __syncthreads();
     }
 
-    if (threadIdx.x < 16) {
-      sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x], sd[threadIdx.x + 16]);
+    if (threadId < 16) {
+      sd[threadId] = RAJA_MINLOC(sd[threadId], sd[threadId + 16]);
     }
     __syncthreads();
 
-    if (threadIdx.x < 8) {
-      sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x], sd[threadIdx.x + 8]);
+    if (threadId < 8) {
+      sd[threadId] = RAJA_MINLOC(sd[threadId], sd[threadId + 8]);
     }
     __syncthreads();
 
-    if (threadIdx.x < 4) {
-      sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x], sd[threadIdx.x + 4]);
+    if (threadId < 4) {
+      sd[threadId] = RAJA_MINLOC(sd[threadId], sd[threadId + 4]);
     }
     __syncthreads();
 
-    if (threadIdx.x < 2) {
-      sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x], sd[threadIdx.x + 2]);
+    if (threadId < 2) {
+      sd[threadId] = RAJA_MINLOC(sd[threadId], sd[threadId + 2]);
     }
     __syncthreads();
 
-    if (threadIdx.x < 1) {
+    if (threadId < 1) {
       lastBlock = false;
-      sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x], sd[threadIdx.x + 1]);
-      m_blockdata[m_blockoffset + blockIdx.x + 1] =
-          RAJA_MINLOC(sd[threadIdx.x],
-                      m_blockdata[m_blockoffset + blockIdx.x + 1]);
+      sd[threadId] = RAJA_MINLOC(sd[threadId], sd[threadId + 1]);
+      m_blockdata[m_blockoffset + blockId + 1] =
+          RAJA_MINLOC(sd[threadId],
+                      m_blockdata[m_blockoffset + blockId + 1]);
+      __threadfence();
```

Comment:
You're absolutely right.  I'm sorry, I shouldn't comment on these things 
when I'm out of it.  The CPU architectures all have it, the GPUs don't.  
I needed one for a structure in OpenCL a couple of years ago and ended 
up implementing one using the exch primitive to build a single-thread 
lock and use a pair of exchanges for write operations.  It's probably 
not worth doing that here.

On 28 Jun 2016, at 16:46, jonesholger wrote:

>>        lastBlock = false;
>> -      sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x], sd[threadIdx.x 
>> + 1]);
>> -      m_blockdata[m_blockoffset + blockIdx.x + 1] =
>> -          RAJA_MINLOC(sd[threadIdx.x],
>> -                      m_blockdata[m_blockoffset + blockIdx.x + 1]);
>> +      sd[threadId] = RAJA_MINLOC(sd[threadId], sd[threadId + 1]);
>> +      m_blockdata[m_blockoffset + blockId + 1] =
>> +          RAJA_MINLOC(sd[threadId],
>> +                      m_blockdata[m_blockoffset + blockId + 1]);
>> +      __threadfence();
>
> Hmm, have to admit to not seeing a 128bit CAS anywhere in the CUDA 7.5 
> docs, but I can dig around for the signature. Also, the compiler also 
> insists that I define an explicit assignment operator for this 
> volatile type .. ughh
>
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/LLNL/RAJA/pull/68/files/e5563ecdd13d0bf3a935d59f26545fee5d9cdf61#r68864014